zorder/src/main/scala/org/locationtech/sfcurve/zorder/ZRange.scala

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2013-2016 Commonwealth Computer Research, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Apache License, Version 2.0 which
+ * accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ */
+
+package org.locationtech.sfcurve.zorder
+
+/**
+ * Represents a rectangle in defined by min and max as two opposing points
+ * 
+ * @param min: lower-left point 
+ * @param max: upper-right point
+ */
+case class ZRange(min: Long, max: Long) {
+
+  require(min <= max, s"Range bounds must be ordered, but $min > $max")
+
+  def mid: Long = (max + min)  >>> 1 // overflow safe mean
+
+  def length: Long = max - min + 1
+
+  // contains in index space (e.g. the long value)
+  def contains(bits: Long): Boolean = bits >= min && bits <= max
+
+  // contains in index space (e.g. the long value)
+  def contains(r: ZRange): Boolean =  contains(r.min) && contains(r.max)
+
+  // overlaps in index space (e.g. the long value)
+  def overlaps(r: ZRange): Boolean = contains(r.min) || contains(r.max)
+}
+
+object ZRange {
+  def apply(min: Z2, max: Z2)(implicit d: DummyImplicit): ZRange = ZRange(min.z, max.z)
```

Comment:
no, b/c Z3 and Z2 are both value classes, the runtime signature for both applies is `apply(Long, Long)`. In order to differentiate the two methods we use the dummy implicits.